New date range selector branch

Boopster opened the pull request.

donal-s commented on reactDashboard/src/components/DataCharts/DataCharts.jsx:
```diff
@@ -72,23 +78,23 @@ export function parseNhsCsvData(csvData) {
     dailyCasesPoints.push({
       t: date,
-      y: cases,
+      y: cases
```
Can you rerun your autoformat on these files please? :)

Boopster replied:
I think this is resolved now

donal-s commented:
There are a lot of additional changes here apart from the date range selection - could you list the things we should be looking out for please?

Boopster commented:
> There are a lot of additional changes here apart from the date range selection - could you list the things we should be looking out for please?

Work that came from additional tickets:- 

1. moved sonification button, added text and improved visibility, 

2. changed all tooltips/popups to black background and left aligned - some previous changes had been overwritten and this resolved

The other tidy up I did here was to remove remnants for full-screen charts - only the button had been removed previously when this feature was dropped.

The rest of it was all in the original ticket I think

donal-s commented:
First run (after npm install) gave the following:

```
TypeError: lines is null
createDateAggregateValuesMap
src/components/Utils/CsvUtils.js:113

  110 | export function createDateAggregateValuesMap(lines) {
  111 |   const result = new Map();
  112 | 
> 113 |   lines.forEach(
  114 |     (
  115 |       [
  116 |         dateString,

getNhsCsvDataDateRange
src/components/Utils/CsvUtils.js:5

  2 | import { differenceInDays, format } from "date-fns";
  3 | 
  4 | export function getNhsCsvDataDateRange(csvDataHB, csvDataCA = null) {
> 5 |   let dateAggregateValuesMap = createDateAggregateValuesMap(csvDataHB);
  6 | 
  7 |   let dates = [...dateAggregateValuesMap.keys()].sort();
  8 | 

Overview/<
src/pages/Overview.jsx:82

  79 | useEffect(() => {
  80 |   // Only attempt to fetch data once
  81 |   if (healthBoardDataset != null || councilAreaDataset != null) {
> 82 |     const parseDateRange = getNhsCsvDataDateRange(
     | ^  83 |       healthBoardDataset,
  84 |       councilAreaDataset
  85 |     );
```

donal-s commented on reactDashboard/src/components/ChartDropdown/ChartDropdown.test.js:
```diff
@@ -0,0 +1,133 @@
+  container = null;
+});
+
+function checkDropdownText() {
```
Could use a couple of testcases which show menu content with and without showPercentageCases set

donal-s commented on reactDashboard/src/components/ChartDropdown/ChartDropdown.jsx:
```diff
@@ -0,0 +1,78 @@
+        value={getChartTypeText(chartType)}
+        title={getChartTypeText(chartType)}
+      >
+        {chartType == null ? "Select a chart" : getChartTypeText(chartType)}
```
Is it valid to ever have chartType null ? It can never be undefined as it defaults to DAILY_CASES above. I'd simplify it by having it throw an error if null is passed in (and have a test case to prove it)

Boopster replied:
I have removed line 51 - I think the logic within getChartTypeText(chartType) captures this and has been tested

donal-s commented on reactDashboard/src/components/DataCharts/DataChartsSelector.jsx:
```diff
@@ -1,61 +0,0 @@
-  ) {
-    throw new Error("Unrecognised chartType: " + chartType);
-  }
-  if (setChartType === null || setChartType === undefined) {
```
This validation logic might be of use in the replacement chart selector (along with the associated test cases)

Boopster replied:
Thanks - will look into it

Boopster replied:
see comment above

donal-s commented on reactDashboard/src/components/DataCharts/DataChartsUtils.js:
```diff
@@ -110,8 +121,7 @@ export function commonChartConfiguration(datasets) {
       legend: {
-        position: "bottom",
-        onClick: (e) => e.stopPropagation(),
+        display: false,
```
While needing a legend is debatable if there is a single trace and its meaning is clear from the chart title (I'd have one anyway), a legend is necessary if there is more than one trace (as in the regional insights chart)

Boopster replied:
Good point - will look into this also

Boopster replied:
![image](https://user-images.githubusercontent.com/22414785/95985805-f6f42280-0e1c-11eb-9d44-c48cd7284a0f.png)

Boopster replied:
back in for both regional and overview

Boopster replied:
![image](https://user-images.githubusercontent.com/22414785/95986269-9fa28200-0e1d-11eb-87f1-51f7a3acc10e.png)

Boopster replied:
adjusted label box width

donal-s commented on reactDashboard/src/pages/Regional.jsx:
```diff
@@ -47,6 +47,23 @@ const Regional = ({
+  });
+
+  useEffect(() => {
+    // Only attempt to fetch data once
```
The comment is no longer accurate - replace with: Get maximum date range to encompass all data (or similar) - or just bin the comment

Boopster replied:
Binned

donal-s commented on reactDashboard/src/pages/Regional.jsx:
```diff
@@ -129,16 +146,15 @@ const Regional = ({
+        <Col xs={12}>
           <RegionDataCharts
             chartType={chartType}
+            setChartType={setChartType}
```
The only reason chartType exists outside the new DataCharts/RegionalDataCharts is to allow stopping of any playing sonification when the user selects a different chart. It may be cleaner to move the stopAudio() effect (line 106) into the dataChart components along with the chartType/setChartType state. That way it's more self contained within the charts, and the regional and overview pages don't need to care about chartType state.

Boopster replied:
moved to separate ticket

donal-s commented on reactDashboard/src/pages/Regional.jsx:
```diff
@@ -47,6 +47,23 @@ const Regional = ({
 
   const currentRegionCode = useRef(regionCode);
   const currentLocation = useRef(match.url);
+  const [dateRange, setDateRange] = useState("dateRange");
```
Same here - dateRange and maxDateRange state (and the effect that sets their initial values) could be pushed down into the dataChart components - they're not needed for anything else

Boopster replied:
moved to separate ticket

donal-s commented on reactDashboard/src/pages/Regional.jsx:
```diff
@@ -129,16 +146,15 @@ const Regional = ({
         </Col>
       </Row>
       <Row className="fullscreen-charts">
```
Minor point - fullscreen-charts isn't really accurate. We can rename at some point

donal-s commented on reactDashboard/src/pages/Overview.jsx:
```diff
@@ -29,10 +29,13 @@ const Overview = ({
-  const [zoomDataCharts, setZoomDataCharts] = useState(false);
+  const [dateRange, setDateRange] = useState({ startDate: 0, endDate: 0 });
   const [zoomGeoMap, setZoomGeoMap] = useState(false);
+  const [maxDateRange, setMaxDateRange] = useState({
```
Same comments about pushing state and logic down into the dataChart component as made in Regional.jsx

Boopster replied:
moved to separate ticket

donal-s commented on reactDashboard/src/components/Utils/Sonification.js:
```diff
@@ -160,20 +180,40 @@ export function calculateMaxDataValue(seriesData = null) {
  * @param {number[]} seriesData - Dataset to sonify.
  * @param {string} [place = "Scotland"] - region name for dataset. If undefined, defaults to 'Scotland'.
```
Missing a parameter in the docs here

Boopster replied:
added: 

 * @param {object} dateRange - startDate and endDate for intro speech.

donal-s commented on reactDashboard/src/components/Utils/CsvUtils.js:
```diff
@@ -1,6 +1,27 @@
 import moment from "moment";
 import { differenceInDays, format } from "date-fns";
 
+export function getNhsCsvDataDateRange(csvDataHB, csvDataCA = null) {
```
Test coverage?

Boopster replied:
Had a go :)

donal-s commented on reactDashboard/src/components/GeoHeatMap/RegionGeoMap.jsx:
```diff
@@ -62,7 +62,7 @@ const RegionGeoMap = ({
       const featureCode = featureCodeForFeature(layer.feature);
       const content =
-        "<strong>" + getPlaceNameByFeatureCode(featureCode) + "</strong>";
+        "<p class='region-popup'><strong>" + getPlaceNameByFeatureCode(featureCode) + "</strong></p>";
```
It looks like there's a blank line under the place names in the tooltips now - can you remove that?

Boopster replied:
![image](https://user-images.githubusercontent.com/22414785/95983573-c2329c00-0e19-11eb-82cf-271dd68bfb34.png)

Boopster replied:
removed

donal-s commented on reactDashboard/src/components/DataCharts/QuickSelectDateRange.jsx:
```diff
@@ -0,0 +1,66 @@
+
+  return (
+    <div className="quick-select-dates">
+      <Button
```
there's a lot of replication of code here - you could create a function createButton(id, timePeriod, buttonText) to reduce that.

Boopster replied:
moved to separate ticket

donal-s commented on reactDashboard/src/components/DataCharts/DateRangeSlider.test.js:
```diff
@@ -0,0 +1,159 @@
+      })
+    ).toStrictEqual(expectedResult);
+  });
+  it("defaultValues", () => {
```
Space out the it() functions with newlines between them please? Makes the test cases easier to distinguish

Boopster replied:
added spaces

donal-s commented on reactDashboard/src/components/DataCharts/DateRangeSlider.jsx:
```diff
@@ -0,0 +1,147 @@
+      boxShadow: "#ccc 0 2px 3px 1px",
+    },
+    "& .bar": {
+      // display: inline-block !important;
```
Remove ?

Boopster replied:
removed

donal-s commented on reactDashboard/src/components/DataCharts/DateRangeSlider.jsx:
```diff
@@ -0,0 +1,147 @@
+import { getNhsCsvDataDateRange } from "../Utils/CsvUtils";
+import "./DataCharts.css";
+
+const useStyles = makeStyles({
```
There seems to be two layers of styling here - can they be combined into one?

Boopster replied:
moved to another ticket

donal-s commented on reactDashboard/src/components/DataCharts/DataCharts.jsx:
```diff
@@ -171,18 +181,6 @@ const DataCharts = ({
         borderColor: "DarkGray;",
         borderWidth: 1,
         position: "nearest",
-        callbacks: {
```
Why was this removed?

Boopster replied:
I thought this just related to the label for the legend - I will re-insert now that is back

donal-s replied:
Nope, tis to format the numbers on the data point tooltips

donal-s commented on reactDashboard/src/components/DataCharts/DataCharts.css:
```diff
@@ -21,19 +63,112 @@
+    top: 20px;
+  }
+  .sonification-play-button h5 {
+    font-size: 0.5rem;
```
I don't think reducing the font size here has much benefit

Boopster replied:
removed

donal-s reviewed (changes requested):
A few comments :)

donal-s commented on reactDashboard/src/pages/Overview.jsx:
```diff
@@ -72,10 +75,21 @@ const Overview = ({
 
+  useEffect(() => {
+    // Only attempt to fetch data once
+    if (healthBoardDataset != null || councilAreaDataset != null) {
```
There's a bug here - if councilAreaDataset is loaded before healthBoardDataset, then getNhsCsvDataDateRange is called with null, councilAreaDataset. I think line 80 should be:

if (healthBoardDataset != null) {

This is a strong argument for adding some test coverage of the pages - even just checking they render.

Boopster replied:
updated line 80

donal-s commented on reactDashboard/src/pages/Regional.jsx:
```diff
@@ -47,6 +47,23 @@ const Regional = ({
+
+  useEffect(() => {
+    // Only attempt to fetch data once
+    if (healthBoardDataset != null || councilAreaDataset != null) {
```
Same issue here

donal-s replied:
The issue here was needing if (healthBoardDataset != null) { to prevent a race condition bug

donal-s commented:
Updates reviewed - remaining unresolved comments pending

donal-s reviewed (commented):
Yup - good progress

donal-s reviewed (approved):
As discussed, there's still some test coverage missing, but agreed to add that to the refactoring Jira issue for now
